packages/bundle-source/src/index.js

```diff
@@ -73,51 +74,79 @@ export default async function bundleSource(
     if (isEntry) {
       entrypoint = fileName;
     }
-    let unmappedCode = code;
+
+    // Parse the rolled-up chunk with Babel.
+    const ast = (babelParser.parse || babelParser)(code);
+
+    let unmapLoc;
     if (
       moduleFormat === 'nestedEvaluate' &&
       !fileName.startsWith('_virtual/')
     ) {
-      // TODO: Should parse the generated chunk with Babel.
-      // We rearrange the generated chunk according to its sourcemap to move
-      // its source lines back to the right place, like Babel generator's
-      // `retainLines: true`.
-      unmappedCode = '';
+      // We rearrange the rolled-up chunk according to its sourcemap to move
+      // its source lines back to the right place.
       // eslint-disable-next-line no-await-in-loop
       const consumer = await new SourceMapConsumer(chunk.map);
-      const genLines = code.split('\n');
-      let lastLine = 1;
-      for (let genLine = 0; genLine < genLines.length; genLine += 1) {
-        const pos = consumer.originalPositionFor({
-          line: genLine + 1,
-          column: 0,
-        });
-        const { line: origLine } = pos;
-
-        const srcLine = origLine === null ? lastLine : origLine;
-        const priorChar = unmappedCode[unmappedCode.length - 1];
-        if (
-          srcLine === lastLine &&
-          !['\n', ';', '{', undefined].includes(priorChar) &&
-          !['}'].includes(genLines[genLine][0])
-        ) {
-          unmappedCode += `;`;
+      const unmapped = new WeakSet();
+      let lastPos = { ...ast.loc.start };
+      unmapLoc = loc => {
+        if (!loc || unmapped.has(loc)) {
+          return;
         }
-        while (lastLine < srcLine) {
-          unmappedCode += `\n`;
-          lastLine += 1;
+        // Make sure things start at least at the right place.
+        loc.end = { ...loc.start };
+        for (const pos of ['start', 'end']) {
+          if (loc[pos]) {
+            const newPos = consumer.originalPositionFor(loc[pos]);
+            if (newPos.source !== null) {
+              lastPos = {
+                line: newPos.line,
+                column: newPos.column,
+              };
+            }
+            loc[pos] = lastPos;
+          }
         }
-        unmappedCode += genLines[genLine];
-      }
+        unmapped.add(loc);
+      };
     }
 
-    // Rewrite apparent import expressions so that they don't fail under SES.
-    // We also do apparent HTML comments.
-    const defangedCode = unmappedCode
-      .replace(IMPORT_RE, '$1notreally')
-      .replace(HTML_COMMENT_RE, '<->');
-    // console.log(`<<<<<< ${fileName}\n${defangedCode}\n>>>>>>>>`);
-    sourceBundle[fileName] = defangedCode;
+    const rewriteComment = node => {
+      node.type = 'CommentBlock';
+      node.value = node.value
+        .replace(/^\s*/gm, '') // Strip extraneous comment whitespace.
+        .replace(HTML_COMMENT_START_RE, '<!X-')
+        .replace(HTML_COMMENT_END_RE, '-X>')
```

Comment:
Could use a comment like “Here we replace markers that resemble HTML comments within JS comments like `// <!--` with `// <!X--` so they can pass through the SES bans on risky source patterns.

Reply:
I can't put those strings in comments or they'll fail under SES. 😝 

I can (and have) described it with prose, though.